[INFINITY-2223] Determine upgradesFrom automatically

elezar opened the pull request:
This PR adds `{{upgrades-from}}` (and `{{downgrades-to}}`) templating to the `UniversePackageBuilder`.

This is done through a simple `PackageManager` wrapper around universe (supporting both `curl` and `requests`). In order to provide the correct sorting of the versions returned by universe, a `Version` class is used which takes both the `releaseVersion` (effectively the folder number in the universe repository) and the `version` (a string version specified by us) into consideration. The latest version is the one with the **highest** `releaseVersion`.

Currently, I have a `VersionResolver` class which processes the output of the `PackageManager.get_packages()` method to determine the available package versions (similar to the `dcos package describe <package-name> --package-versions` command, and also provides the latest version.

In order to integrate this more cleanly into the `UniversePackageBuilder`, I also introduced a `Package` class which encapsulates the package specifics.

In the case where the latest version cannot be determined, `{{upgrades-from}}` (and `{{downgrades-to}}`) is set to `"*"` so that all upgrades are allowed.

# TODO:
- [x] Add templating to other frameworks
- [x] Ensure that `requests` times out too

nickbp commented on tools/universe_builder.py:
```diff
@@ -1,237 +1,43 @@
+    parser.add_argument('package_name', type=str, help='The package name')
+    parser.add_argument('package_version', type=str, help='The package version string')
+    parser.add_argument('package_dir_path', type=str, help='The local path where the package template is located')
+    parser.add_argument('upload_dir_url', type=str, help='url of the directory where artifacts are located (S3, etc):')
```
Is the trailing colon in the help string intentional?

nickbp commented on tools/universe/package.py:
```diff
@@ -0,0 +1,34 @@
+        self._name = name
+        self._version = version
+
+        self._set_version_resolver(version_resolver)
```
Having this broken out in a separate function feels a little overkill, would it make sense to have the contained logic directly under `__init__`?

nickbp commented on tools/universe/package.py:
```diff
@@ -0,0 +1,34 @@
+
+        self._set_version_resolver(version_resolver)
+
+    def _set_version_resolver(self, version_resolver):
```
Python likes double newlines between functions (here and below)

elezar replied:
As far as I am aware, PEP8 only requires the double newlines between module-level functions or classes, for class methods, or other whitespace, a single blank line is required. I fill run `flake8` on the files here again to be sure.

nickbp commented on tools/universe/package_builder.py:
```diff
@@ -0,0 +1,210 @@
+            self._artifact_file_paths[os.path.basename(
+                artifact_path)] = artifact_path
+
+    def set_input_dir_path(self, input_dir_path):
```
IIRC, python likes double newlines between functions (here and below). Looks like the previous version didn't have them, but may as well fix it while here...

nickbp commented on tools/universe/package_manager.py:
```diff
@@ -0,0 +1,87 @@
+
+        self.__package_cache = None
+
+    def get_packages(self):
```
IIRC, python likes double newlines between functions (here and below). Looks like the previous version didn't have them, but may as well fix it while here...

nickbp commented on tools/universe/version_resolver.py:
```diff
@@ -0,0 +1,63 @@
+        return str(self.package_version)
+
+
+class VersionResolver:
```
This feels pretty tied to the `PackageManager` implementation, in terms of expecting the Universe response data to match a certain format. It may make things a little simpler to just combine the two into a single class that does any version extraction internally

nickbp commented on tools/universe_builder.py:
```diff
@@ -1,237 +1,43 @@
-#!/usr/bin/env python3
```
Given that this has a main() function it may make sense to retain the executable bit + crunchbang so that it's directly executable. IIRC some of the bash tooling may call directly into `universe_builder.py`.

nickbp commented on tools/universe/package.py:
```diff
@@ -0,0 +1,34 @@
+            self._version_resolver = version_resolver
+
+    def get_upgrades_from(self):
+        if self._version_resolver is None:
```
Given the logic in _set_version_resolver(), wouldn't this be unreachable?

nickbp commented on tools/universe/package.py:
```diff
@@ -0,0 +1,34 @@
+from .version_resolver import VersionResolver
+
+
+class Package:
```
Now that I think about it, how about moving the version lookup outside of the Package class, and having the Package class function as a data-only struct/container for a name+version?

- `PackageManager`/`VersionResolver` could accept a package name as a parameter and return a Package object.
- `_get_template_mapping_for_content()` could deal with lookup and `"*"` fallback internally
- `PackageManager`'s cache would be more useful since it would survive across package names by default. It could hold a pre-parsed dict of `name => Package` too. Debug logs could print the list of package versions that it found by dumping that dict.

(This comment was written after the following comments)

nickbp replied:
Also, I could see us wanting to have deeper relationships between `Package` objects at some point, for example having a `Package` object graph that follows the `upgradesFrom`/`downgradesTo` paths. Having `PackageManager` construct the `Package`s directly could make constructing this graph a little easier.

benclarkwood commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Determine upgradeFrom](https://trello.com/c/wa4TRfim/26-determine-upgradefrom)

elezar commented:
Thanks for the review @nickbp. You made some good points. I have made some changes, most notably the following:

* `VersionResolver` and `PackageManager` are no longer separate classes, but a single `PackageManager` class.
* The `PackageManger` methods now return `Package` objects.
* The package cache is now a dictionary keyed on the package name with values all the versioned `Package` objects.
* The handling of the fallback to `"*"` is handled in `UniversePackageBuilder`
* The formatting of the files has been addressed.

nickbp commented on tools/universe/package.py:
```diff
@@ -0,0 +1,82 @@
+            'version': str(self.get_version())
+         })
+
+    # def get_upgrades_from(self):
```
remove dead code imo

nickbp commented on tools/universe/test_package.py:
```diff
@@ -0,0 +1,56 @@
+
+
+# def test_no_version_resolver_upgrades_from_all(mocker):
+
```
dead code

nickbp commented on tools/universe_builder.py:
```diff
@@ -1,237 +1,62 @@
-        '  Example: $ {} kafka 1.2.3-4.5.6 /path/to/template/jsons/ https://example.com/path/to/kafka-artifacts /path/to/artifact1.zip /path/to/artifact2.zip /path/to/artifact3.zip'.format(argv[0]))
-    logger.info(
-        'In addition, environment variables named \'TEMPLATE_SOME_PARAMETER\' will be inserted against the provided package template (with params of the form \'{{some-parameter}}\')')
+EPILOG_STRING = "In addition, environment variables named 'TEMPLATE_SOME_PARAMETER' " \
```
`EPILOGUE`?

elezar replied:
Was basing the name on the fact that `argparse` refers to this as the `epilog`: https://docs.python.org/3/library/argparse.html#epilog, but will change it.

nickbp commented on tools/universe/package_manager.py:
```diff
@@ -0,0 +1,127 @@
+            status_code = int(output)
+
+            if status_code != 200:
+                raise Exception("Curl returned status code %s", status_code)
```
the constructor to Exception doesn't appear to handle formatting like this automatically (tried manually on py3.5)

nickbp commented on tools/universe/package_builder.py:
```diff
@@ -0,0 +1,236 @@
+        jsonpath = os.path.join(
+            scratchdir,
+            'stub-universe-{}.json'.format(self._package.get_name()))
+        jsonfile = open(jsonpath, 'w')
```
while we're here... this could be `with open(jsonpath, 'w') as jsonfile: ...`

nickbp replied:
looks like this is still the same?

elezar commented:
Thanks @nickbp. I think I addressed your comments.

There was also an issue with the way we sort versions that was causing elastic to fail, but this should also have been addressed.

elezar commented:
Reverted a change that must have included the json file fix. Will reapply.

elezar commented:
The passing of the HDFS tests would require that #1619 be merged.
